webrtc-c/canary/src/CanaryWebrtc.cpp

```diff
@@ -149,48 +172,20 @@ STATUS onNewConnection(Canary::PPeer pPeer)
     return retStatus;
 }
 
-VOID sendLocalFrames(Canary::PPeer pPeer, MEDIA_STREAM_TRACK_KIND kind, const std::string& pattern, UINT64 frameCount, UINT32 frameDuration)
+VOID sendCustomFrames(Canary::PPeer pPeer, MEDIA_STREAM_TRACK_KIND kind, UINT64 dataRate, UINT64 frameRate)
 {
     STATUS retStatus = STATUS_SUCCESS;
     Frame frame;
-    UINT64 fileIndex = 0, frameSize;
-    CHAR filePath[MAX_PATH_LEN + 1];
-    UINT64 startTime, lastFrameTime, elapsed;
-
-    frame.frameData = NULL;
-    frame.size = 0;
-    frame.presentationTs = 0;
-    startTime = GETTIME();
-    lastFrameTime = startTime;
-
+    frame.size = CANARY_METADATA_SIZE + (dataRate / frameRate);
```

Comment:
Recommend having a new structure that wraps/encapsulates Frame like so

typedef struct {
    UINT64 field1;
    UINT64 field2;
...
} CanaryFrameMetadata;

typedef struct {
    Frame frame;
    CanaryFrameMetadata canaryMetadata;
    // -> actual bits of the frame following the matadata;
} CanaryFrame, *PCanaryFrame;

The allocation would happen something like this

PCanaryFrame pCanaryFrame = (PCanaryFrame) MEMALLOC(SIZEOF(CanaryFrame) + frameSize);
CHK(pCanaryFrame != NULL, STATUS_NOT_ENOUGH_MEMORY);

// Set the fields up
pCanaryFrame->frame.frameData = (PBYTE)(pCanaryFrame + 1);
pCanaryFrame->frame.frameSize = frameSize + SIZEOF(CanaryFrameMetadata);
...

Reply:
This will work directly if both sides are on the same machine with the same packing and endianness. 
Care must be taken to handle packing and endianness.

For now, using direct packing into the buffer is a better option I think